src/message/fetchActions.js

```diff
@@ -222,6 +227,42 @@ const fetchTopMostNarrow = () => async (dispatch: Dispatch, getState: GetState)
   }
 };
 
+/**
+ * Calls an async function and if unsuccessful retries the call.
+ *
+ * If the function is an API call and the response has HTTP status code 4xx
+ * the error is considered unrecoverable and the exception is rethrown, to be
+ * handled further up in the call stack.
+ */
+export async function tryFetch<T>(func: () => Promise<T>): Promise<T> {
+  const MAX_TIME_MS: number = 60000;
+  const backoffMachine = new BackoffMachine();
+
+  const timeAtStart = Date.now();
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const msElapsed = Date.now() - timeAtStart;
+    // If the first attempt had a 60-second timeout, and it took 3
+    // seconds, then the second attempt should have a 57-second
+    // timeout. Etc.
+    const timeLimitMs = MAX_TIME_MS - msElapsed;
+    try {
+      return await promiseTimeout<T>(func(), timeLimitMs);
+    } catch (e) {
+      if (isClientError(e) || e instanceof TimeoutError) {
+        throw e;
+      }
+      await backoffMachine.wait();
+    }
+  }
```

Comment:
Looking at this and particularly with the comment I just made about the `backoffMachine.wait()`, I wonder: with the spiffy new `promiseTimeout` helper, could all this timeout logic be simplified by using that around the whole thing? I think it could.

Reply:
Hmm, if I understand correctly, something like this?

```javascript
export async function tryFetch<T>(func: () => Promise<T>): Promise<T> {
  const MAX_TIME_MS: number = 60000;
  const backoffMachine = new BackoffMachine();

  return promiseTimeout<T>(
    (async () => {
      // eslint-disable-next-line no-constant-condition
      while (true) {
        try {
          return await func();
        } catch (e) {
          if (isClientError(e) || e instanceof TimeoutError) {
            throw e;
          }
          await backoffMachine.wait();
        }
      }
      // Without this, Flow 0.92.1 does not know this code is unreachable,
      // and it incorrectly thinks Promise<undefined> could be returned,
      // which is inconsistent with the stated Promise<T> return type.
      // eslint-disable-next-line no-unreachable
      throw new Error();
    })(),
    MAX_TIME_MS,
  );
}
```

Reply:
Oh, hmm, no, that wouldn't do it—the infinite loop could continue infinitely even after the timer was up. I'll see what I can think of.

Reply:
Er, maybe this isn't the most elegant solution, but it might work?

```javascript
export async function tryFetch<T>(func: () => Promise<T>): Promise<T> {
  const MAX_TIME_MS: number = 60000;
  const backoffMachine = new BackoffMachine();

  let timerHasExpired = false;

  return promiseTimeout(
    (async () => {
      // eslint-disable-next-line no-constant-condition
      while (true) {
        if (timerHasExpired) {
          // No one is listening for this Promise's outcome, so stop
          // doing more work.
          throw new Error();
        }
        try {
          return await func();
        } catch (e) {
          if (isClientError(e)) {
            throw e;
          }
          await backoffMachine.wait();
        }
      }
      // Without this, Flow 0.92.1 does not know this code is unreachable,
      // and it incorrectly thinks Promise<undefined> could be returned,
      // which is inconsistent with the stated Promise<T> return type.
      // eslint-disable-next-line no-unreachable
      throw new Error();
    })(),
    MAX_TIME_MS,
    () => {
      timerHasExpired = true;
      throw new TimeoutError();
    }
  );
}
```